design/design-doc-package-refactor.md

```diff
@@ -0,0 +1,690 @@
+# Package Refactor
+* Owner: Dan Mangum (@hasheddan)
+* Reviewers: Crossplane Maintainers
+* Status: Draft
+
+## Background
+
+Crossplane currently supports installing controllers and new CRDs into a
+Kubernetes cluster that the Crossplane package manager is running in. While
+there are many other packaging formats in the Kubernetes ecosystem, Crossplane
+supports its own for the following reasons:
+
+- Crossplane is [opinionated about the capabilities of a
+  controller](https://github.com/crossplane/crossplane/blob/master/design/one-pager-packages-security-isolation.md#allowed-resource-access)
+  that can be installed to extend its functionality. For instance, controllers
+  [may not run as
+  root](https://github.com/crossplane/crossplane/blob/master/design/one-pager-packages-security-isolation.md#package-deployment-privileges)
+  or request cluster admin RBAC.
+- Crossplane [allocates and aggregates various
+  ClusterRoles](https://github.com/crossplane/crossplane/blob/master/design/one-pager-packages-security-isolation.md#crossplane-clusterroles--rbac)
+  to automatically provide permissions for users in the Kubernetes cluster to
+  view / edit / create / delete CRDs installed by a package.
+- Crossplane guards against conflicting CRDs being installed into a cluster.
+- Crossplane adds [additional metadata to
+  CRDs](https://github.com/crossplane/crossplane/blob/master/design/one-pager-stack-ui-metadata.md#crd-annotation-example)
+  to provide additional context for displaying their configuration in a UI.
+- Crossplane [adds labels to
+  CRDs](https://github.com/crossplane/crossplane/blob/master/design/one-pager-stack-relationship-labels.md#example-wordpress-crdyaml-parented-by-stackinstall)
+  in an attempt to establish parent-child relationships between CRDs.
+
+In addition, the following unimplemented features are goals of the Crossplane
+package manager:
+
+- The ability to resolve dependencies listed in a package for CRDs that do not
+  yet exist in the cluster.
+- The ability for a user to upgrade installed package versions without manual
+  intervention.
```

Comment:
No, it is not in scope for Crossplane to automatically upgrade packages without a user modifying the image on a `Package`. How do you feel about saying the following instead of "manual intervention":

> The ability for a user to upgrade installed package versions without manually managing the package artifacts.

Reply:
That seems clearer. Or perhaps something like:

> The ability for a use to upgrade installed package artifacts simply by updating the Package CR?